### PR TITLE
POC improve JMAP message destruction speed

### PR DESCRIPTION
**Pull request by chibenwa**
Asa bonus, we will also avoid failing on large GetMessages calls

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 1]
```diff
@@ -162,23 +165,69 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
...
-        ImmutableList<MetadataWithMailboxId> metadataWithMailbox = messageIdMapper
+        List<MailboxMessage> messageList = messageIdMapper
             .find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata)
             .stream()
             .filter(inMailboxes(mailboxIds))
+            .collect(Guavate.toImmutableList());
+
+        if (!messageList.isEmpty()) {
+            delete(messageIdMapper, messageList, mailboxIds, mailboxSession);
+            return DeleteResult.builder()
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: you should add static methods for simple cases like `DeleteResult.destroyed(messageId)` or `DeteleResult.notFound(messageId)`

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 2]
```diff
@@ -162,23 +165,69 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
...
+    @Override
+    public DeleteResult delete(List<MessageId> messageIds, MailboxSession mailboxSession) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+
+        List<MailboxMessage> messageList = messageIdMapper.find(messageIds, MessageMapper.FetchType.Metadata);
+        ImmutableSet<MailboxId> allowedMailboxIds = getAllowedMailboxIds(mailboxSession, messageList, Right.DeleteMessages);
+
+        ImmutableSet<MessageId> accessibleMessages = messageList.stream()
+            .filter(message -> allowedMailboxIds.contains(message.getMailboxId()))
+            .collect(Guavate.toImmutableListMultimap(
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: why do you collect to a multimap if you only need keys in a Set ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 3]
```diff
@@ -162,23 +165,69 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
...
+                    Message::getMessageId,
+                    Function.identity()))
+            .keySet();
+        Sets.SetView<MessageId> nonAccessibleMessages = Sets.difference(ImmutableSet.copyOf(messageIds), accessibleMessages);
+
+        if (!accessibleMessages.isEmpty()) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: why do we need this if ? calling delete with an empty list should be supported

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 4]
```diff
@@ -162,23 +165,69 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
...
+            delete(messageIdMapper, messageList, allowedMailboxIds, mailboxSession);
+        }
+
+        return DeleteResult.builder()
+            .addDestroyed(accessibleMessages)
+            .addNotFound(nonAccessibleMessages)
+            .build();
+    }
+
+    public void delete(MessageIdMapper messageIdMapper, List<MailboxMessage> messageList, Collection<MailboxId> inMailboxes, MailboxSession mailboxSession) throws MailboxException {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: private ?

Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: inMailboxes looks unused

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 5]
```diff
@@ -101,10 +102,13 @@ public void setFlagsShouldNotFailWhenMessageDoesNotExist() throws Exception {
     }
 
     @Test
-    public void deleteMessageShouldNotFailWhenMessageDoesNotExist() throws Exception {
+    public void deleteMessageShouldThrowWhenMessageDoesNotExist() throws MailboxException {
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: it doesn't throw, it reports a failure in a DeleteResult

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java [hunk 6]
```diff
@@ -310,6 +333,16 @@ public void deleteMessageShouldThrowExceptionWhenDeletingOnOtherSession() throws
         messageIdManager.delete(messageId, ImmutableList.of(aliceMailbox1.getMailboxId()), bobSession);
     }
 
+    @Test
+    public void deleteAllShouldThrowExceptionWhenDeletingOnOtherSession() throws Exception {
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerStorageTest.java]: doesn't throw neither

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/1433

**Comment by chibenwa** [comment]
Refined version merged

**Pull request closed without merging**